aten/src/ATen/native/RNN.cpp

```diff
@@ -0,0 +1,691 @@
+    for (size_t i = 0; i < params.size(); i += 2) {
+      result.emplace_back(params[i], params[i + 1], undefined, undefined);
+    }
+  }
+  return result;
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+// HIDDEN STATE FUNCTIONS
+//
+// Functions implemented below are implemented as templates based on hidden type,
+// because they need to work both with simple RNNs and GRU (which use a single Tensor),
+// as well as with LSTM (or possibly more complicated architectures in the future).
+// Still, there are some operations that need to be performed on the hidden states
+// alone, and for this purpose we provide an overloaded set of functions below.
+
+Tensor hidden_as_output(const Tensor& t) { return t; }
+Tensor hidden_as_output(const tpair_of<Tensor>& t) { return std::get<0>(t); }
+
+template<size_t index>
+std::vector<Tensor> project(at::ArrayRef<tpair_of<Tensor>> tuples) {
+  std::vector<Tensor> result;
+  result.reserve(tuples.size());
+  for (auto & t : tuples) {
+    result.push_back(std::get<index>(t));
+  }
+  return result;
+}
+
+Tensor hidden_concat(at::ArrayRef<Tensor> hiddens) { return at::cat(hiddens, 0); }
+tpair_of<Tensor> hidden_concat(at::ArrayRef<tpair_of<Tensor>> hiddens) {
+  return std::make_tuple(hidden_concat(project<0>(hiddens)), hidden_concat(project<1>(hiddens)));
+}
+
+Tensor hidden_slice(const Tensor& t, int64_t start, int64_t end) {
+  return t.narrow(0, start, end - start);
+}
+tpair_of<Tensor> hidden_slice(const tpair_of<Tensor>& t, int64_t start, int64_t end) {
+  return std::make_tuple(hidden_slice(std::get<0>(t), start, end),
+                         hidden_slice(std::get<1>(t), start, end));
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// CELL IMPLEMENTATIONS
+//
+// Cell is a basic component of an RNN, representing a single application of the
+// recurrent function. You can think of it as a function of signature
+//
+// (Tensor input, hidden_type hidden, CellParams) -> hidden_type
+//
+// which means that it consumes an input tensor, and updates the previous hidden state.
+// It's a struct only because functional programming in C++ is a pain, and it's easier
+// to pass around "vtable pointers" than actual function pointers.
+
+Tensor linear(const Tensor& input, const Tensor& weight, /* optional */ const Tensor& bias={}) {
+  if (input.dim() == 2 && bias.defined()) {
+    // fused op is marginally faster
+    return at::addmm(bias, input, weight.t());
+  }
+
+  auto output = at::matmul(input, weight.t());
+  if (bias.defined()) {
+    output.add_(bias);
+  }
+  return output;
+}
+
+template<typename hidden_type_tmpl>
+struct Cell {
+  using hidden_type = hidden_type_tmpl;
+  virtual hidden_type operator()(const Tensor& input, const hidden_type& hidden, const CellParams& params) const = 0;
+};
+
+template<Tensor (*nonlinearity)(const Tensor&)>
```

Comment:
nit: Give a name for the type of a nonlinearity function, it's easier to understand than the anonymous syntax for function pointer (blugh).